Sprint 14

iamjohansson opened the pull request:
Сверстан главный экран с двумя items Трекеры и Статистика (во второй добавлен только заголовок и плейсхолдер)
На экране трекера реализовано отображение созданных трекеров (с возможностью их отмечать и вести статистику в кол-ве дней), дэйтпикер, окно поиска и кнопка добавления нового трека.
При переходе через + получаем окно с выбором, какой тип трекера нам создать.
В соответствии с выбранным типом, переходим на экран настройки трекера.
После успешной настройки, тапаем соответствующую кнопку и переходим обратно в основное окно со списком трекеров.

iamjohansson committed:
Первичная настройка приложения:
Сделан лаунчскрин, сверстан таббар с 2 итемами (Окном трекера и статистики)
Добавлена смена сцены на ТБконтроллер
Добавлено расширение для UIColor

iamjohansson committed:
Сверстаны кнопка и лейбл, цвета поддерживают светлую/темную тему

iamjohansson committed:
Добавлены структуры для UI

iamjohansson committed:
Добавлена UI часть ячейки для трекера;
Добавлен экран создания трекера

iamjohansson committed:
Добавлена UI часть для настройки трекера

iamjohansson committed:
Доработка UI части Сеттинг контроллера

iamjohansson committed:
Сделана отдельная настройка ТекстФилда, в соответствии с макетом
Добавлен Рекогнайзер, для скрытия клавиатуры, при тапе по области

iamjohansson committed:
Добавлена ячейка и UI часть блока с расписанием

iamjohansson committed:
Исправлены ошибки, поправлен код.
Реализован переход на Расписание и возврат с него, с упорядоченными шорт названиями в дескрипшин лейбле
Добавлены цвета для треков и смайлы, пока будут принимать случайное значение

iamjohansson committed:
Реализована финальная логика, создаются привычки и нерегулярки

iamjohansson committed:
Финальные правки кода + чистка + структурирование
Добавлена тень у таббара, плейсхолдер вынесен в расширение в котором конфигурируется, доверстан блок статистики

iamjohansson committed:
Заблокированы будущие даты, пользуемся только текущей или прошедшими.

helecta commented on Tracker/Cells/ScheduleCell.swift:
```diff
@@ -0,0 +1,85 @@
+    }()
+    
+    // MARK: - Properties
+    weak var delegate: ScheduleCellDelegate?
```
Стоит организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Желательно пройтись по всему проекту и поправить код

iamjohansson replied:
ок обязательно пройдусь

helecta commented on Tracker/Helpers/UIColor + Extension.swift:
```diff
@@ -0,0 +1,34 @@
+
+extension UIColor {
+    // MARK: - System Colors - for light and dark theme
+    static var Background: UIColor { UIColor(named: "Background") ?? UIColor.darkGray }
```
Такие константы стоит называть со строчной буквы. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

iamjohansson replied:
+

helecta commented on Tracker/Helpers/Int + Extension.swift:
```diff
@@ -0,0 +1,30 @@
+extension Int {
+    
+    // MARK: - Word Declension Method
+        func daysString() -> String {
```
Есть еще один способ получше, но для этого нужно будет добавить локализацию для приложения. Но в локализации может быть один язык, зато в будущем сильно облегчит перевод приложения на другие языки.
[Общая статья про локализацию приложения](https://habr.com/ru/post/419077/)
И [статья про локализацию существительных множественного числа](https://crunchybagel.com/localizing-plurals-in-ios-development/)

iamjohansson committed:
Доработка:
Можно выбрать будущую дату, но отметить трекер нельзя.
В место этого пользователь получит алерт, в котором указано, что сегодня (текущее число) и завершить трекер на будущую дату нельзя.

iamjohansson commented:
В целом можно и заблокировать кнопку полностью, но с уведомлением вроде симпатичнее.

iamjohansson committed:
Поправил наименование переменных системного цвета

iamjohansson committed:
Микроправка текста алерта

iamjohansson committed:
Визуально, при выборе будущей даты, кнопка становится более прозрачной

iamjohansson commented:
Добавил визуальное отображение, что кнопка не кликабельна.
Блок кода isEnable в ячейке закомменчен, если требуется полный блок, можно включить.

iamjohansson committed:
чистка кода
